crates/runtime/src/flight/flightsql/prepared_statement_query.rs

```diff
@@ -102,19 +124,153 @@ pub(crate) async fn do_get(
 
     let datafusion = Arc::clone(&flight_svc.datafusion);
     tracing::trace!("do_get: {query:?}");
-    match std::str::from_utf8(&query.prepared_statement_handle) {
-        Ok(sql) => {
-            let (output, from_cache) =
-                Box::pin(Service::sql_to_flight_stream(datafusion, sql)).await?;
-            let timed_output = TimedStream::new(output, move || start);
-
-            let mut response =
-                Response::new(Box::pin(timed_output) as <Service as FlightService>::DoGetStream);
-            attach_cache_metadata(&mut response, from_cache);
-            Ok(response)
+
+    let handle =
+        String::from_utf8(query.prepared_statement_handle.to_vec()).map_err(error_to_status)?;
+
+    let (sql, parameters) = match &flight_svc.prepared_statements.read().await.get(&handle) {
+        Some(PreparedStatement { query, parameters }) => (query.clone(), parameters.clone()),
+        None => (handle, None),
+    };
+
+    let (output, from_cache) =
+        Box::pin(Service::sql_to_flight_stream(datafusion, &sql, parameters)).await?;
+    let timed_output = TimedStream::new(output, move || start);
+
+    let mut response =
+        Response::new(Box::pin(timed_output) as <Service as FlightService>::DoGetStream);
+    attach_cache_metadata(&mut response, from_cache);
+    Ok(response)
+}
+
+pub(crate) async fn do_put_query(
+    flight_svc: &Service,
+    query: CommandPreparedStatementQuery,
+    streaming_flight: Peekable<Streaming<FlightData>>,
+) -> Result<Response<<Service as FlightService>::DoPutStream>, Status> {
+    let streaming_flight = streaming_flight
+        .map(|flight_data| flight_data.map_err(|e| arrow_flight::error::FlightError::Tonic(e)));
+
+    let mut decoder = FlightDataDecoder::new(streaming_flight);
+    let schema = decode_schema(&mut decoder).await?;
+
+    let mut parameters = Vec::new();
+    let mut encoder = StreamWriter::try_new(&mut parameters, &schema).map_err(error_to_status)?;
+    let mut total_rows = 0;
+    while let Some(msg) = decoder.try_next().await? {
+        match msg.payload {
+            DecodedPayload::None => {}
+            DecodedPayload::Schema(_) => {
+                return Err(Status::invalid_argument(
+                    "parameter flight data must contain a single schema",
+                ));
+            }
+            DecodedPayload::RecordBatch(record_batch) => {
+                total_rows += record_batch.num_rows();
+                encoder.write(&record_batch).map_err(error_to_status)?;
+            }
         }
-        Err(e) => Err(Status::invalid_argument(format!(
-            "Invalid prepared statement handle: {e}"
-        ))),
     }
+    if total_rows > 1 {
+        return Err(Status::invalid_argument(
+            "parameters should contain a single row",
+        ));
+    }
+
+    let parameters = if parameters.is_empty() {
+        None
+    } else {
+        Some(parameters.as_bytes())
```

Comment:
nit: Why not call decode_param_values in here and have decode_param_values be:

fn decode_param_values(
    parameters: &[u8],
)

instead of 

fn decode_param_values(
    parameters: Option<&[u8]>,
)